approvaltests/utils.py

```diff
@@ -17,12 +17,17 @@ def write_to_temporary_file(expected, name):
         return temp.name
 
 
-def to_json(object):
-    return json.dumps(
-        object,
-        sort_keys=True,
-        indent=4,
-        separators=(',', ': '),
-        default=lambda o: o.__dict__,
-        ensure_ascii=True,
-    )
\ No newline at end of file
+def to_json(obj):
+    try:
+        return json.dumps(
+            obj,
+            sort_keys=True,
+            indent=4,
+            separators=(',', ': '),
+            default=lambda o: o.__dict__,
+            ensure_ascii=True,
+        )
+    except AttributeError as e:
```

Comment:
What input `obj` would trigger this? From the tests I thought that `Ellipsis` would but I get a `TypeError`:

```
>>> import sys; print(sys.version_info)
sys.version_info(major=3, minor=7, micro=5, releaselevel='final', serial=0)
>>> import json
>>> json.dumps(Ellipsis)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type ellipsis is not JSON serializable
```

Reply:
Good question! One example would be a class which uses `__slots__`, which prevents `__dict__` being created. I suppose in future this could be accommodated by interrogating `__slots__` too.

    >>> class NoDict:
    ...     __slots__ = ["a"]
    ... 
    >>> obj = NoDict()
    >>> obj.a = 42
    >>> obj.a
    42
    >>> import json
    >>> json.dumps(obj, sort_keys=True, indent=4, separators=(',', ': '), default=lambda o: o.__dict__, ensure_ascii=True)
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/usr/local/var/pyenv/versions/3.8.0/lib/python3.8/json/__init__.py", line 234, in dumps
    return cls(
      File "/usr/local/var/pyenv/versions/3.8.0/lib/python3.8/json/encoder.py", line 201, in encode
    chunks = list(chunks)
      File "/usr/local/var/pyenv/versions/3.8.0/lib/python3.8/json/encoder.py", line 438, in _iterencode
    o = _default(o)
      File "<stdin>", line 1, in <lambda>
    AttributeError: 'NoDict' object has no attribute '__dict__'

Reply:
Understood, thanks! I think it would be useful to mention slots as an example in the comment. Also, is there a test for this branch? If we think a comment and test would be useful, it could come through a separate PR as this one's been held up long enough. Let me know what you think.

Reply:
I'll add some more tests and comments in the next day or two. Better support for __slots__ can follow later.